### ISSUE TITLE
Segmentation fault when fine-tuning stt_fa_fastconformer_hybrid_large due to Numba/RNNT loss

### ISSUE THREAD

**Issue by KiarashQ** [NVIDIA-NeMo/NeMo#14986]
Hello NeMo Team,

I am trying to fine-tune the nvidia/stt_fa_fastconformer_hybrid_large model on a custom Farsi dataset, but I am consistently encountering a fatal Segmentation fault (or silent kernel death in Jupyter) during the first training step.

The pre-trained model has a very high WER (46%) on my test set, so fine-tuning is critical.

**The Problem:**

The crash happens immediately after the trainer.fit() call, right after the initial validation sanity check and before the first training batch completes. The final log message before the crash is always a warning related to the Numba RNNT loss calculation.

**Environment:**

- OS: Ubuntu 22.04 (in a Docker container)
- GPU: NVIDIA GeForce RTX 4070 SUPER
- Python: 3.10 (via Conda)
- Key Libraries (Installed in a clean Conda env):
  - pytorch: 2.3.1+cu121
  - pytorch-lightning: 1.9.5
  - nemo_toolkit[asr]: 1.21.0
  - numba: 0.59.1

**Summary of Extensive Debugging:**

We have systematically ruled out all common causes for this type of crash:

1. Out of Memory: The crash occurs even with batch_size=1 and on a tiny 20-sample smoke test dataset. nvidia-smi shows low VRAM and power usage at the moment of the crash.
2. Hardware/Power Issues: Power limiting the GPU (nvidia-smi -pl 220) has no effect. A minimal PyTorch stress test (allocating large tensors and performing matrix multiplication) passes without any issues, proving the hardware, driver, and core PyTorch installation are stable.
3. Environment Issues: The issue persists after a complete environment rebuild using Conda with pinned, known-compatible library versions.
4. Calling Method: The crash occurs identically whether using an interactive, object-oriented approach in a Jupyter Notebook or using the official examples/asr/speech_to_text_finetune.py script.
5. Dataloader: A direct test of the dataloader (batch = next(iter(train_dataloader))) passes successfully. The crash only happens when the Trainer brings the model and data together for the training_step.

**The Conclusive Finding:**

The final traceback before the crash points directly to the rnnt_loss_gpu function, which is implemented with Numba.
```
Current thread 0x00007caf781b1740 (most recent call first):
  File ".../numba/cuda/cudadrv/driver.py", line 326 in safe_cuda_api_call
  ...
  File ".../nemo/collections/asr/parts/numba/rnnt_loss/rnnt.py", line 174 in rnnt_loss_gpu
  ...
  File ".../nemo/collections/asr/models/hybrid_rnnt_ctc_models.py", line 425 in training_step
```

We have also confirmed that even setting cfg.loss.rnnt_weight = 0.0 does not prevent this function from being called, leading to the same crash.
This strongly suggests an intractable, low-level bug within the warprnnt_numba library when used with this specific model in this environment.

**The Ask:**

1. Is this a known issue?
2. Is there a "golden configuration" of NVIDIA drivers, CUDA, and library versions that is known to work for fine-tuning this Farsi Hybrid model?
3. Is there any way to truly disable the RNNT loss and fine-tune this model using only its CTC head? The model appears to be hard-coded to always call the RNNT loss function.
4. Thank you for your time and for the fantastic work on NeMo. Any guidance would be deeply appreciated.

**Comment by KiarashQ** [comment]
## Update

I was able to resolve the segmentation fault that occurred during the first training step (inside `rnnt_loss_gpu` from `nemo/collections/asr/parts/numba/rnnt_loss/rnnt.py`) when fine-tuning nvidia/stt_fa_fastconformer_hybrid_large.

### Root Cause

The crash was triggered by the Numba-based RNNT loss kernel, which was still being called even when `cfg.loss.rnnt_weight = 0.0`. On Ada GPUs (e.g., RTX 4070 SUPER with CUDA 12.1), this path consistently causes a low-level segfault under FP16.

### Fix

I replaced the built-in RNNT loss call with a custom hybrid loss function that conditionally skips RNNT and computes only the CTC loss when desired.
This prevents `rnnt_loss_gpu` from being invoked at all.

In addition, for stability I also:

- switched to `precision="bf16-mixed"`
- limited to `devices=1`
- set all dataloader `num_workers = 0`

### Example
```python
# Custom loss (simplified)
def compute_loss(batch, model):
    feats, feats_len, tokens, tokens_len = batch
    logits, logit_len, _ = model.forward(input_signal=feats, input_signal_length=feats_len)
    ctc_loss = model.loss_ctc(logits, logit_len, tokens, tokens_len)
    return ctc_loss 

trainer = L.Trainer(accelerator="gpu", devices=1, precision="bf16-mixed")

asr_model.cfg.train_ds.num_workers = 0
asr_model.cfg.validation_ds.num_workers = 0
```

### However: how to use the Numba RNNT loss correctly?

Could you advise on a known-good configuration for the Numba RNNT loss on Ada + CUDA 12.x? Specifically:

- Versions
- Backend selection
- Runtime flags/practices

**Comment by parthchadha** [comment]
@artbataev @nithinraok can you please help answer the question about good config for RNNT?
